### PR TITLE
feat(rome_js_formatter): TS Intersection & Union types #3162

### LINKED ISSUE #3162: Formatter: TS Intersection & Union types

### PR DESCRIPTION
**Pull request by denbezrukov**
## Summary

Fix #3162 for TS Union.

Prettier: https://github.com/prettier/prettier/blob/cd3e530c2e51fb8296c0fb7738a9afdd3a3a4410/src/language-js/print/type-annotation.js#L123-L202

## Test Plan

Current:
**Average compatibility**: 86.00
**Compatible lines**: 84.91

Main:
**Average compatibility**: 85.67
**Compatible lines**: 84.09

New test cases `union_type.ts`
`cargo test --package rome_js_formatter`

### CODE REVIEW

**Review by MichaReiser — APPROVED** [review]
Awesome. I'll have some fun merging this with my comments refactor :D

File: crates/rome_js_formatter/src/ts/types/union_type.rs [hunk 1]
```diff
@@ -1,67 +1,110 @@
...
+                        soft_line_break(),
+                        if_group_breaks(&text(")"))
+                    ]
                 )
-            )
-        };
+            } else if should_indent {
+                write!(f, [&indent(&types)])
+            } else {
+                write!(f, [&types])
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/ts/types/union_type.rs:103]: Nit: You don't need to pass references to `write` and `format_args`
```suggestion
                write!(f, [indent(&types)])
            } else {
                write!(f, [types])
```

**Review by MichaReiser — APPROVED** [review]

File: crates/rome_js_formatter/src/ts/types/union_type.rs (hunk 1 above)
Comment (denbezrukov) [reply crates/rome_js_formatter/src/ts/types/union_type.rs]: fixed (=

**Comment by denbezrukov** [comment]
> Awesome. I'll have some fun merging this with my comments refactor :D

Oh, sorry.
We can wait your MR and after I can try to merge it after your refactor. Is it better?

**Comment by MichaReiser** [comment]
> > Awesome. I'll have some fun merging this with my comments refactor :D
> 
> Oh, sorry. We can wait your MR and after I can try to merge it after your refactor. Is it better?
> 
> BTW: I've found strange comments behaviour with union types. [Playground](https://play.rome.tools/?code=type+A+%3D+%5B%0A++++%2F*leading+comment+with+new+line*%2F+%0A++++A+%7C+B%2C%0A++++%5D%3B%0A%0Atype+A+%3D+%0A++%2F*leading+comment+with+new+line*%2F+%0A++++A+%7C+B%3B&lineWidth=80&indentStyle=tab&quoteStyle=double&quoteProperties=as-needed&indentWidth=2&sourceType=module&typescript=true&jsx=false#dAB5AHAAZQAgAEEAIAA9ACAAWwAKACAAIAAgACAALwAqAGwAZQBhAGQAaQBuAGcAIABjAG8AbQBtAGUAbgB0ACAAdwBpAHQAaAAgAG4AZQB3ACAAbABpAG4AZQAqAC8AIAAKACAAIAAgACAAQQAgAHwAIABCACwACgAgACAAIAAgAF0AOwAKAAoAdAB5AHAAZQAgAEEAIAA9ACAACgAgACAALwAqAGwAZQBhAGQAaQBuAGcAIABjAG8AbQBtAGUAbgB0ACAAdwBpAHQAaAAgAG4AZQB3ACAAbABpAG4AZQAqAC8AIAAKACAAIAAgACAAQQAgAHwAIABCADsA)

Don't be. It helps me a big deal when working on matching the union and intersection type comments handling because I now only have too look into how comments are formatted and can assume that the rest works as expected. 

Don't worry. I'll look into the comment placement. We soon have better tooling to place comments well.

**Pull request merged**